modules/internal/ChapelIO.chpl

```diff
@@ -17,36 +17,310 @@
+
+When programming the input and output method for a custom data type, it is
+often useful to define both the read and write routines at the same time. That
+is possible to do in a Chapel program by defining a ``readWriteThis`` method,
+which is a generic method expecting a single argument: either a Reader or a
+Writer.
+
+In cases when the reading routine and the writing routine are more naturally
+separate, or in which only one should be defined, a Chapel program can define
+``readThis`` (taking in a single argument of type Reader) and/or
+``writeThis`` (taking in a single argument of type Writer).
+
+If a none of these routines are provided, a default version of ``readThis`` and
+``writeThis`` will be generated by the compiler. If ``readWriteThis`` is
+defined, the compiler will generate ``readThis`` or ``writeThis`` - if they do
+not already exist - which call ``readWriteThis``.
+
+Objects of type class:`Reader` or class:`Writer` both support the following
+fields and methods:
+
+ * :attr:`Reader.writing` :attr:`Writer.writing` returns `true` if called on a class:``Writer`` and `false` if called on a class:``Reader``.
+ * :attr:`Reader.binary` :attr:`Writer.binary` returns `true` if called on a Reader or Writer that is configured to perform binary I/O, and `false` otherwise.
+ * :proc:`Reader.error` :proc:`Writer.error` returns a saved error code.
+ * :proc:`Reader.setError` :proc:`Writer.setError` save an error code.
+ * :proc:`Reader.clearError` :proc:`Writer.clearError` clear any saved error code.
+ * :proc:`Reader.readwrite` :proc:`Writer.readwrite` read or write a value according to its readThis or writeThis method.
+ * :proc:`Reader.readWriteLiteral` :proc:`Writer.readWriteLiteral` read or write a literal string value.
+ * :proc:`Reader.readWriteNewline` :proc:`Writer.readWriteNewline` reads and discards input until a newline or writes a newline.
+
+Objects of type :class:`Reader` also supports the following methods:
+
+ * :proc:`Reader.read` (similar to :proc:`IO.channel.read`)
+ * :proc:`Reader.readln` (similar to :proc:`IO.channel.readln`)
+
+Objects of type :class:`Writer` also support the following methods:
+
+ * :proc:`Writer.write` (similar to :proc:`IO.channel.read`)
+ * :proc:`Writer.writeln` (similar to :proc:`IO.channel.writeln`)
+
+Note that objects of type :class:`Reader` or :class:`Writer` may represent a
+locked channel; as a result, using parallelism constructs to call methods on
+:class:`Reader` or :class:`Writer` may result in undefined behavior.
+
+.. note::
+
+  In the future, we plan to merge :class:`Reader`, :class:`Writer`, and
+  :record:`IO.channel` in order to support calling :proc:`IO.channel.readf`
+  and other I/O routines inside a ``readThis`` method.
+
+Because it is often more convenient to use an operator for I/O, instead of
+writing
+
+.. code-block:: chapel
+
+  f.readwrite(x);
+  f.readwrite(y);
+
+one can write
+
+.. code-block:: chapel
+
+  f <~> x <~> y;
```

Comment:
Can you take an action item to add it?  It seems like something that should live there.